### ISSUE TITLE
If app appears on non-active screen while active screen has screen scaling enabled, window resizes

### ISSUE THREAD

**Issue by wyhinton** [fltk/fltk#259]
Moved this issue from [https://github.com/fltk-rs/fltk-rs/issues/784](https://github.com/fltk-rs/fltk-rs/issues/784) to here, as it seems to be purely related to C FLTK. 

When I launch my FLTK app, sometimes the window automatically resizes. This behavior only occurs when the cursor is not on the same screen as the app when the app window appears. That is, the app launches on the "non-active" window.  

This resizing can also sometimes cause widgets to resize strangely.

I'm using the same monitor setup as when we were working through the gidplus issue. (launching the app on the active screen in VSCode, then having the app appear on the "non active" monitor). [https://github.com/fltk-rs/fltk-rs/issues/740](https://github.com/fltk-rs/fltk-rs/issues/740).  Monitor 0  is 1920x1080 and monitor 1 is 4k. 

Os is Win10x64
```C
#include <FL/Fl.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Window.H>

int main() {
    auto win = new Fl_Window(500, 500, "Counter");
    auto but1 = new Fl_Button(100, 100, 100, 100, "Dummy");
    auto but2 = new Fl_Box(200, 100, 100, 100, "Dummy");
    but1->tooltip("Tooltip..");
    // auto but3 = new FL_Button
    win->end();
    win->show();
    Fl::lock();
    return Fl::run(); 
}
```
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.15)
project(monitors)

include(FetchContent)

set(FLTK_BUILD_TEST OFF CACHE BOOL " " FORCE)
set(OPTION_USE_GDIPLUS ON CACHE BOOL " " FORCE)

FetchContent_Declare(FLTK
    GIT_REPOSITORY https://github.com/fltk/fltk
    GIT_SHALLOW TRUE
)
FetchContent_MakeAvailable(FLTK)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE fltk gdiplus)
target_include_directories(main PRIVATE ${fltk_SOURCE_DIR} ${fltk_BINARY_DIR})
target_link_libraries(main PRIVATE fltk)

```
build.bat
```
cmake -S. -Bbin -DCMAKE_BUILD_TYPE=Release && cmake --build bin --config Release
```
Window resizing itself when it appears on the non-active monitor:
![resize_mon_issue](https://user-images.githubusercontent.com/38958118/126080889-e8df231c-886c-423a-9e07-4ca93507d9fe.gif)

**Comment by Albrecht-S** [comment]
> launching the app on the active screen in VSCode, then having the app appear on the "non active" monitor
> Monitor 0 is 1920x1080 and monitor 1 is 4k.

You missed to tell us:
- which one of these do you call the "active" monitor/screen running VSCode
- which screen zoom factors did you set in the Windows monitor setup

I could try to look this up in the other issue you mentioned (which is pretty long), but for others reading this here, please provide that information.

**Comment by wyhinton** [comment]
> You missed to tell us:
> which one of these do you call the "active" monitor/screen running VSCode
> which screen zoom factors did you set in the Windows monitor setup
> 

> which one of these do you call the "active" monitor/screen running VSCode

The "active" screen, as I understand it with how FLTK works on windows, is the one the mouse is located on. Moving the mouse to the monitor makes it the active screen. The bug occurs when you launch the application on the active screen with screen scaling enabled (be it vscode, command prompt, file explorer), and then the FLTK window appears on the non-active screen. 

> which screen zoom factors did you set in the Windows monitor setup

I had it set to 175% when I made this issue, but the bug persists with every screen scale setting except 100%.

Changing the screen scaling on the 4k monitor to 100%, we no longer get the issue:
![image](https://user-images.githubusercontent.com/38958118/126081435-a8ef8b65-d93b-41e0-8f4b-2a971aca7a44.png)

**Comment by Albrecht-S** [comment]
Let's put the "launching the program and window resizing" effect aside for a moment.

What happens if you move the window from one screen to the other screen and back? The expected behavior would be that the window keeps its size, as far as possible, if ... Assume the screen **size** is similar (e.g. 21") but the screens have different pixel sizes (effective: different resolutions, dpi). If you set the screen scaling factors in a way that they have similar **effective** **resolutions**, then the perceived window size should be "the same" on both monitors. In this case moving the window from one monitor to the other monitor would - at some point in the move - change the internal scaling factor to make the window resize and appear to have (about) the same size.

I don't recall the physical sizes of your monitors. But assume they are the same size (21") and standing side by side. The left (2k = 1920x1080) monitor has scaling factor 100%. The right (4k = 3840x2160) monitor is at scaling factor 200% (which reduces its effective resolution to 1920x1080). Then the window should keep its approximate **physical size** when you move it from one monitor to the other.

In your setup this might work pretty well if the 2k monitor is @100% and the 4k monitor is @200% or another appropriate scaling factor such that the resolution after applying the scaling is **physically** the same.

Why am I describing all this? When you move the window from one monitor to the other one there is (should be) a small delay until the window changes its scaling factor (this is done for a "better user experience"). I can't test this, but while you're moving the window from the 2k to the 4k monitor it may first appear bigger (on the 4k monitor) while moving but then - after a short delay - change its size to the "previous" size because the scaling factor is adjusted. This - or something like this - seems to be what you're seeing when you launch the program on the 4k monitor and it gets displayed on the 2k monitor. I'm asking you to test this and report what you see (when moving the window between monitors) because I can't do this myself.

**Comment by wyhinton** [comment]
> What happens if you move the window from one screen to the other screen and back? The expected behavior would be that the window keeps its size, as far as possible, if ... Assume the screen size is similar (e.g. 21") but the screens have different pixel sizes (effective: different resolutions, dpi). 
Yes the screen sizes are similar,  21" for the 1080 and 24" for the 4k, I believe. 

Yes, there is a small delay in the resizing. I've always understood that FLTK was doing some resizing of the window when moving it between mointors with different resolutions. I guess the problem is that in this particular situation where the window is popping up on the "non-active" window causes the window to appear in the incorrect position and can cause widgets to take on the wrong dimensions, as I described in the fltk-rs issue.

Screen capture: 1080p monitor on the left, 4k on the right:
https://user-images.githubusercontent.com/38958118/126364402-24674a70-df0f-4bcb-9677-ae3c6a77a75f.mp4

**Comment by Albrecht-S** [comment]
Thanks for testing and for the video.

**Comment by rminderhoud** [comment]
I can confirm this happens for me as well, my repro steps are like so:

- Start FLTK app on main window at 1080p resolution with 150% scaling. Screen flashes briefly due to some quick resizing
- Drag the window to second monitor at 1080p resolution with 100% scaling, window resizes briefly

Some side effects: This triggers sub windows to now have a border which was previously not there.

**Comment by MatthiasWM** [comment]
I remember that Windows 10 had huge issues with HiDPI screens in general. For every application, there was a setting on how to handle HiDPI screens. Depending on the user setting, Windows would sometimes scale windows, sometimes not, and sometimes scale the windows, but not the font. A lot of software was unusable. I am pretty sure that 3 years later, a lot changed, but I am afraid that the OS is still fighting against FLTK to find the correct window scaling.

As mentioned here:

https://learn.microsoft.com/en-us/windows/win32/hidpi/high-dpi-desktop-application-development-on-windows

[WM_DPICHANGED](https://learn.microsoft.com/en-us/windows/win32/hidpi/wm-dpichanged) and [WM_GETDPISCALEDSIZE](https://learn.microsoft.com/en-us/windows/win32/hidpi/wm-getdpiscaledsize) may not be treated correctly. "Windows can, however, forcibly change the DPI awareness mode...". 

I am not a Windows developer. It would be nice if anyone could fix this after such a long time....

**Comment by MatthiasWM** [comment]
I do not have a Windows machine available right now, but I think I found the bug. The line `1898` in `Fl_win32.cxx` read 
```
        UINT dpi = UINT(sd->dpi[screen_num()][0]);
```
using the current mouse position to calculate the window DPI before opening it. Once the window is open, it will receive a WM_DPICHANGED and the do a delayed resize to the correct DPI.

The fix should be easy:
```
        UINT dpi = UINT(sd->dpi[screen_num(r.left, r.top)][0]);
```
so it recognizes the DPI of the top left of the the window that will be opened.

Things to consider:
- [ ] does the patch fix things?
- [ ] what happens if screen_num() can;t find a screen?
- [ ] should the coordinates be the top left or rather the center of the window

If someone could test that and answer my questions, I can write a patch and commit it.

**Comment by MatthiasWM** [comment]
If I don;t hear anything, I will try harder to get a Windows PC with two screens and apply my patch. 

Expires **PATCH FIRST** on Dec 23rd.

**Comment by ManoloFLTK** [comment]
I can run my UTM VM with 2-display Windows 11 and will look at this issue.

**Comment by ManoloFLTK** [comment]
This issue should be fixed with the commit mentioned above.

**Comment by MatthiasWM** [comment]
@ManoloFLTK thank you very much for testing. I should have time to push this tomorrow.

**Comment by MatthiasWM** [comment]
Oh, ok, just saw the code. Makes sense.

**Comment by MatthiasWM** [comment]
Extended to Jan 10th to get confirmation

**Comment by ManoloFLTK** [comment]
Closing because confirmation did not arrive before set expiration date.
